test/integration/test_tracing.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+from __future__ import annotations
+
+import json
+import logging
+import time
+from typing import Any, Callable
+
+import httpx
+import jubilant
+
+
+def test_direct_connection(build_charm: Callable[[], str], juju: jubilant.Juju):
+    charm_path = build_charm()
+    juju.deploy(charm_path)
+    juju.integrate('tracing-tester', 'tempo')
+    status = juju.wait(jubilant.all_active)
+
+    spans = wait_spans(status.apps['tempo-worker'].address, test=lambda s: 'ops.main' in str(s))
+    assert 'ops.main' in [s['name'] for s in spans]
+
+    event_names = [e['name'] for e in get_events(spans)]
+    assert 'StartEvent' in event_names
+
+    checkpoint = time.time()
```

Comment:
Does this require the clocks being in sync? Seems likely given these are ephemeral VMs for CI, but I wonder if we can avoid by setting `since` to the timestamp of the last span received?

Reply:
Yes this requires clocks to be in sync; they should be between the test host and the k8s that runs on the test host.

There's no guarantee that we've gotten all spans, so I wouldn't want to depend on the "start of last span that we did get".

In this specific test, it's not a very big deal, because no actions are run before this point, however, I kindof like the idiom of scoping the spans we are looking at.

A full set of traces and spans is more than a screenful even for a simple test like this. I'd rather help whoever may need to debug this if/when there's a potential regression.

Reply:
Yep, sounds good.